### ISSUE TITLE
has prepared statement

### ISSUE THREAD

**Issue by emicklei** [jackc/pgx#94]
In order to use the cached prepared statements of a connection, I need to conditionally create them.

```
if !conn.HasPreparedStatement("myname") {
    conn.Prepare("myname", "myquery")
}
_, err = conn.Exec("myname", args)
```

or should i use a different pattern here?  I noticed that it is calling Prepare too much because sendSimpleQuery associates each new prepared statement under ""  (why not use the sql as key?)

**Comment by emicklei** [comment]
With

```
func (c *Conn) HasPreparedStatement(name string) bool {
	_, ok := c.preparedStatements[name]
	return ok
}
```
I can use
```
if !conn.HasPreparedStatement("EnqueueMessage") {
	conn.Prepare("EnqueueMessage", "select publish_message($1,$2,$3,$4,$5,$6);")
}
_, err := conn.Exec("EnqueueMessage",
	topic,
	msg.ContentType,
	msg.ContentEncoding,
	msg.ContentMD5,
	msg.GzipPayload,
	props)

```

btw, using sql as the key is a bad idea.

**Comment by jackc** [comment]
Can you create all your prepared statements ahead of time? That is what the AfterConnect hook is for.

As far as Prepare("") -- it is special. The PostgreSQL wire protocol recognizes the unnamed prepared statement as a special temporary prepared statement.

**Comment by emicklei** [comment]
yes, creating them at connect is possible but complicates maintainance a
bit. The query definition is no longer close the code location where
parameters are set for execution

On Friday, September 18, 2015, Jack Christensen <notifications@github.com>
wrote:

> Can you create all your prepared statements ahead of time? That is what
> the AfterConnect hook is for.
>
> As far as Prepare("") -- it is special. The PostgreSQL wire protocol
> recognizes the unnamed prepared statement as a special temporary prepared
> statement.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/jackc/pgx/issues/94#issuecomment-141264441>.
>


-- 
^ Ernest Micklei

http://ernestmicklei.com

**Comment by emicklei** [comment]
Would you accept a MR for this?

**Comment by jackc** [comment]
I want to have this use case improved, but this issue relates to a few other concerns. So I'd like to discuss these issues and how they relate to this before we commit to a path.

It sounds like you want to acquire a connection, prepare a statement unless it already has been prepared, execute the query, then release the connection back into the queue. Is the only reason for this to optionally prepare the query? i.e. If this wasn't an issue, could you just call ConnPool.Exec and lose the boiler-plate of connection management? I find that I typically only explicitly check out connections for listen/notify. If this is the case then this issue strongly relates to #68.

Acquiring a connection, preparing a statement, and releasing the connection without deallocating the statement bothers me a little. It means that we have changed the persistent state of the connection such that when we acquire a connection from the pool it will not be identical to other connections in the pool. I don't see a concrete case where it would cause a problem, but I'd want to give it a close look to be sure.

As far as what you originally requested, I'm open to exposing the current collection of prepared statements. Though I think I would prefer a GetPreparedStatement that followed the the "comma, ok" pattern. With that you can easily check for existence, but you can also inspect the prepared statement data if desired.

Another approach that might be even cleaner would be to make `Prepare` idempotent. If the name and SQL already match an existing prepared statement it becomes a no-op. This would let you avoid the if statement and simply Prepare and Exec. This also avoids increasing the public interface of Conn.

Yet another approach that could solve the issue is automatic prepared statement creation and caching. This could be a bit too automagical, but it would be nice to remove the need to deal with prepared statements manually altogether.

So to summarize we have several approaches some of which interrelate:

* GetPreparedStatement
* Prepare becomes idempotent
* Add prepared statements to ConnPool
* Automatic prepared statement caching

Any thoughts?

**Comment by emicklei** [comment]
Thank you for opening this discussion. This way we can find the correct solution.

My usecase is indeed that I use a connection pool (max 30) where each member will need to call one of the storedprocedures (~20). After doing some pgx log inspection, I noticed the prepare calls and then realized I had to explicity call Prepare once per connection. That's how I came to the HasPrepare.

You mentioned deallocating statements which I overlooked to be honest. Maybe it is because I can't think of a problematic situation either. For my use case, all connections in the same pool are used for the procedure calls.

I like the proposal for an idempotent Prepare. Like I mentioned before, I prefer the have the Prepare calls close the actual use of the statement (for clarity and maintenance) instead of putting them all in the AfterConnect function.

Prepared statements on the ConnPool level may sounds like a better approach, but I agree with you  on #68  that is difficult to get right and therefore problably not worth adding the required complexity.

**Comment by jackc** [comment]
I added idempotent Prepare.

I still plan on looking into automatic prepared statement creation and caching down the road -- we'll see where that leads.
